### ISSUE TITLE
[Accelerate] Each Training step takes an absurd amount of time

### ISSUE THREAD

**Issue by neel04** [huggingface/accelerate#257]
Training a VQ-VAE-2 on Kaggle's GPU - when I start training on a full epoch, each iteration (i.e each step/pass on a single batch) takes around 5 mins on GPU. GPU usage is negligible, yet it does occupy all VRAM as CUDA should 🤷 

This is my training function:-
```py
def train(loader, val_loader, scheduler):
    accelerator = Accelerator(fp16=False, cpu=args.cpu_run)
    device = accelerator.device

    #initializing the model
    model = VQVAE(in_channel=3, channel=128, n_res_block=args.res_blocks,
                  n_res_channel=args.res_channel,
                  embed_dim=args.embed_dim, n_embed=args.n_embed,
                  decay=args.decay).to(device)

    optimizer = optim.Adam(model.parameters(), lr=args.lr)

    accelerator.print(summary(model, (args.batch_size, 3, args.size, args.size)))
    accelerator.print(model) #vanilla pytorch summary

    model, optimizer, loader, val_loader = accelerator.prepare(model, optimizer, loader, val_loader)
    loader, val_loader = tqdm(loader), tqdm(val_loader)

    criterion = nn.MSELoss()

    latent_loss_weight = 0.25
    latent_loss_beta_list = torch.linspace(0, latent_loss_weight, 20)

    sample_size = 20

    mse_sum = 0
    mse_n = 0
    val_mse_sum, val_mse_n, beta_index = 0, 0, 0 #init beta index

    with wandb.init(project=args.wandb_project_name, config=args.__dict__, save_code=True, name=args.run_name, magic=True): 
        for epoch in range(args.epoch):
            #Starting Epoch loops
            model.train()
            for i, (img, label) in enumerate(loader):
                img = img.to(device)

                out, latent_loss = model(img)
                recon_loss = criterion(out, img)
                latent_loss = latent_loss.mean()

                beta_index = epoch  #for consistency
                if beta_index >= 24:
                    beta_index = latent_loss_beta_list[-1]

                loss = recon_loss + latent_loss_beta_list[beta_index] * latent_loss

                model.zero_grad(set_to_none=True)

                accelerator.backward(loss) #added loss to backprop

                if scheduler is not None:
                    scheduler.step()

                torch.nn.utils.clip_grad_norm_(model.parameters(), args.gradclip) #grad clipping

                optimizer.step()

                mse_sum += recon_loss.item() * img.shape[0]
                mse_n += img.shape[0]

                lr = optimizer.param_groups[0]['lr']

                wandb.log({"epoch": epoch+1, "mse": recon_loss.item(), 
                            "latent_loss": latent_loss.item(), "avg_mse": (mse_sum / mse_n), 
                            "lr": lr})

                if i % 25 == 0:
                    accelerator.print({"epoch": epoch+1, "mse": recon_loss.item(),
                        "latent_loss": latent_loss.item(), "avg_mse": (mse_sum/ mse_n), 
                        "lr": lr})

                #Performing Validation and loggign out images
                if epoch % 2 == 0:   #i % 100 == 0
                    model.eval()

                    #--------------VALIDATION------------------
                    for i, (img, label) in enumerate(val_loader):
                        img.to(device)
                        model.to(device)

                        with torch.no_grad():
                            out, latent_loss = model(img)

                        val_recon_loss = criterion(out, img)
                        val_latent_loss = latent_loss.mean()
                        val_loss = recon_loss + latent_loss_beta_list[beta_index] * latent_loss

                        val_mse_sum += recon_loss.item() * img.shape[0]
                        val_mse_n += img.shape[0]

                    wandb.log({"epoch": epoch+1, "val_mse": val_recon_loss.item(), 
                            "val_latent_loss": val_latent_loss.item(), "val_avg_mse": (val_mse_sum/ val_mse_n), 
                            "lr": lr})

                    accelerator.print({"epoch": epoch+1, "val_mse": val_recon_loss.item(), 
                            "val_latent_loss": val_latent_loss.item(), "val_avg_mse": (val_mse_sum/ val_mse_n), 
                            "lr": lr})

                    model.train()

            #Saving the model checkpoints every epoch
            if epoch % 10 == 0:
                model.eval()
                
                sample = img[:sample_size]

                with torch.no_grad():
                    out, _ = model(sample)

                utils.save_image(
                    torch.cat([sample, out], 0),
                    f'/kaggle/working/samples/{str(epoch + 1).zfill(5)}_{str(i).zfill(5)}.png',
                    nrow=sample_size,
                    normalize=True,
                    range=(-1, 1),
                )

                wandb.log({f"{epoch+1}_Samples" : [wandb.Image(img) for img in torch.cat( [sample, out], 0) ]})
                accelerator.save(model.state_dict(), f'checkpoint/vqvae_{str(epoch + 1).zfill(3)}.pt')
                model.train()

            print(f'\n---EPOCH {epoch} CCOMPLETED---\n')
```
I don't see how tensor shapes are changing at all - so things should work relatively smoother. 

Validation on around 15,000 images is relatively fine I suppose, 
```py
 98%|████████████████████████████████████████▎| 116/118 [07:37<00:04,  2.48s/it]
```
Though I'd expected it to be much, much faster.

But training, 
```py
2%|▌                                 | 31/1783 [2:37:49<145:33:55, 299.11s/it]
{'epoch': 1, 'val_mse': 0.07451071590185165, 'val_latent_loss': 0.03670966625213623, 'val_avg_mse': 0.13713628306023537, 'lr': 0.0001}
```
is pretty slow. Any ideas why **computations** are being offloaded to CPU while the graph is clearly being built on GPU?

Also,
```py
parser.add_argument('--cpu-run', type=bool, default=False)
```
my CPU run arg is set to `False` by default, so at least I am not overriding GPU computation 🤷‍♂️

**Comment by sgugger** [comment]
It seems you're running on one GPU only, so what speed does the same code run without using `Accelerate`?

**Comment by neel04** [comment]
A non-Accelerate vanilla implementation on a single `P100` performs nearly 10 epochs ~ 12 hours (no validation) which roughly comes about ~2.42 seconds/step on average. I suppose it's using CPU here, but that doesn't make sense if the graph is being built on GPU and CUDA reserves all GPU memory too?

**Comment by sgugger** [comment]
Weird that it changes so much when it's kind of the same. Random though, is `torch_xla` installed in the environment event when not using TPUs?

**Comment by neel04** [comment]
Nope. Can confirm a ModuleNotFoundError when importing

**Comment by neel04** [comment]
I will try printing the `device` variable to confirm whether its indeed using GPU - or is there some bug with the argparsing but apart from that, I don't really see what else would affect computations like that 🤷

**Comment by neel04** [comment]
```py
USING DEVICE: cuda		ARGS @ CPU_RUN: False
```
Definitely using CPU, just computations aren't being run on it

**Comment by sgugger** [comment]
If you're not using mixed precision or DeepSpeed, `accelerator.backward(loss)` is exactly like doing `loss.backward()`. Could you try replacing the lines and see if the slowdown persists?

To check if it's a data bottleneck cause by Accelerate, could you try to not send the `train_dataloader` to the `Accelerator.prepare` method and see if it fixes the training speed?

**Comment by neel04** [comment]
Removed dataloader from `prepare`, some tiny differences in speed but nothing major - GPU is still idle most of the time.

Same with `loss.backward()` - not really any tangible impacts. The whole thing is a mystery; why using `backward` slows down `Accelerate` soo drastically is beyond me. 🤷‍♂️ 

I suppose I can send you the repo+colab in some time, if I am not able to figure out the problem?

**Comment by sgugger** [comment]
Sure.
Could you also print `accelerator.state` to make sure there is not some weird thing going on?

**Comment by neel04** [comment]
```py
Accelerator State: Distributed environment: NO
Num processes: 1
Process index: 0
Local process index: 0
Device: cuda
Use FP16 precision: False
```
I am currently going to revamp my entire training loop by making the `train` function train only for a single epoch, using another external function with a loop to interface and run the `train` and seeing whether that improves anything - because apart from a bug in `Accelerate` the only thing I can think of is my loop 🤔

**Comment by neel04** [comment]
Revamped my training loop, for reference its something like this:-
```py
def exec(epochs, loader, val_loader, model, optimizer, scheduler, device):
  for i in range(epochs):
    train(i, loader, val_loader, model, optimizer, scheduler, device)
    torch.save(model.state_dict(), f'checkpoint/vqvae_{str(i + 1).zfill(3)}.pt')
```
where I use `notebook_launcher(exec)` to launch the training. Very weirdly, training works fine **IFF** validation snippet is removed:-
```py
              if epoch > 0 and epoch % 2 == 0:
                  model.eval()

                  #--------------VALIDATION------------------
                  for i, (img, label) in enumerate(val_loader):
                      img = img.to(device)

                      with torch.no_grad():
                          out, latent_loss = model(img)

                      val_recon_loss = criterion(out, img)
                      val_latent_loss = latent_loss.mean()

                      val_mse_sum += recon_loss.item() * img.shape[0]
                      val_mse_n += img.shape[0]

                  print({"epoch": epoch+1, "val_mse": val_recon_loss.item(), 
                          "val_latent_loss": val_latent_loss.item(), "val_avg_mse": (val_mse_sum/ val_mse_n), 
                          "lr": lr})

                  model.train()

```
I have absolutely no idea why training becomes so slow if I validate - which is concerning because I really don't want to overfit

**Comment by sgugger** [comment]
Uh, I can't say I have seen something like this before. It's really weird! Is the validation dataloader passed to `accelerator.prepare`? Also is `model.train()` called before the `train` function? Might be a side effect of the last line otherwise.

**Comment by neel04** [comment]
Hmm... It seems I might have a clue to the situation. for the validation loop, the first thing it triggers is validation due to modulo `%`
```py
if epoch % 2 == 0:

#WHICH CAN BE FIXED SIMPLY BY
if epoch > 0 and epoch % 2 == 0:
```
What I would guess is that perhaps, when the first few steps are taken `Accelerate` is performing some sort of optimizations and compilations of the computation graph....? But for some reason, validating right, in the beginning, breaks something essential for `Accelerate` leading to broken ops for backproppping 🤷‍♂️ 

Pretty sure it's wrong, but I suppose you get the general idea 😉 

> BTW as an aside, have you ever encountered `free() invalid size` when training on TPU? seems like a memory problem, but RAM usage is barely 10% when the error kicks in and execution stops. I may be doing something wrong on my end though (maybe `autocast` is not supposed to be there for TPU?) so might have to check....

**Comment by sgugger** [comment]
That is very very weird. There is no reason validating first would change anything, especially in a non-distributed env. With no distributed training and mixed precision deactivated, Accelerate really only calls the regular PyTorch lines of the training loop, so there is no reason for any difference with PyTorch :-( 

For TPUs there shouldn't be any `autocast` in your code, this is not supported.

**Comment by neel04** [comment]
I was trying to test it for more than a few epochs, but ran out of GPU time and am trying to use TPU - but have this weird error:-
```py

  0%|                                                  | 0/3565 [00:00<?, ?it/s]
  0%|                                                   | 0/235 [00:00<?, ?it/s]free(): invalid size
```
This is my [code BTW](https://www.toptal.com/developers/hastebin/eduroparaj.py) if you might wanna take a look. 

AFAIK running TPU should be as simple as installing the client and running the script 🤷 Its a `TPUv3_8` from kaggle so has more than enough memory 🤔 any ideas?

**Comment by sgugger** [comment]
I have never encountered that error. Are you sure the model is supported on TPU? Some models do operations that are not supported by PyTorch XLA.

**Comment by neel04** [comment]
it's just a VQ-VAE, so the Conv Layers should be supported in any case (perhaps the quantization process mightn't be) - but I don't see why problems can arise as various other similar models have been run on TPUs in the past 🤷 

As for the validation, I can confirm not computing it early on leads to a normal training loop (tested on about 4 epochs) so perhaps one could put some sort of a reminder/disclaimer in the ReadMe to advise people accidentally doing this?
